src/tty/windows/conpty.rs

```diff
@@ -0,0 +1,279 @@
+            })
+        }
+    };
+}
+
+pub struct Conpty {
+    pub handle: HPCON
+}
+
+impl Drop for Conpty {
+    fn drop(&mut self) {
+        // The pseusdoconsole might already have been closed by the console process exiting.
+        // ClosePseudoConsole will fail with error code 1 in that case.
+        //
+        // This check should be sufficient to avoid that.
+        if !process_should_exit() {
+            let api = CONPTY_API.as_ref().unwrap();
+            let result = unsafe { (api.ClosePseudoConsole)(self.handle) };
+
+            // As noted above, if the pseudoconsole is already closed then
+            // ClosePseudoConsole will fail with the error code 1.
+            // (This was not in the MSDN docs as of Nov 2018.)
+            //
+            // If ClosePseudoConsole is successful then result is S_OK.
+            assert!(result == S_OK);
+        }
+    }
+}
+
+// The Conpty API can be accessed from multiple threads.
+unsafe impl Send for Conpty {}
+unsafe impl Sync for Conpty {}
+
+pub fn new<'a>(
+    config: &Config,
+    options: &Options,
+    size: &SizeInfo,
+    _window_id: Option<usize>,
+) -> Option<Pty<'a>> {
+
+    if !config.enable_experimental_conpty_backend() {
+        return None
+    }
+
+    let api = CONPTY_API.as_ref()?;
+
+    let mut pty_handle = 0 as HPCON;
+
+    // Passing 0 as the size parameter allows the "system default" buffer
+    // size to be used. There may be small performance and memory advantages
+    // to be gained by tuning this in the future, but it's likely a reasonable
+    // start point.
+    let (conout, conout_pty_handle) = miow::pipe::anonymous(0).unwrap();
+    let (conin_pty_handle, conin) = miow::pipe::anonymous(0).unwrap();
+
+    let coord = coord_from_sizeinfo(size)
+        .expect("Overflow when creating initial size on pseudoconsole");
+
+    // Create the Pseudo Console, using the pipes
+    let result = unsafe { (api.CreatePseudoConsole)(
+        coord,
+        conin_pty_handle.into_raw_handle(),
+        conout_pty_handle.into_raw_handle(),
+        0,
+        &mut pty_handle as *mut HPCON
+    ) };
+
+    assert!(result == S_OK);
+
+    let mut success;
+
+    // Prepare child process startup info
+
+    let mut size: SIZE_T = 0;
+
+    let mut startup_info_ex: STARTUPINFOEXW = Default::default();
+    startup_info_ex.StartupInfo.cb = mem::size_of::<STARTUPINFOEXW>() as u32;
+
+    // Create the appropriately sized thread attribute list.
+    unsafe {
+        success = InitializeProcThreadAttributeList(
+            ptr::null_mut(),
+            1,
+            0,
+            &mut size as PSIZE_T
+        ) > 0;
+
+        // This call was expected to return false.
+        assert!(!success);
+    }
+
+    let mut attr_list: Box<[BYTE]> = vec![0; size].into_boxed_slice();
+
+    // Set startup info's attribute list & initialize it
+    startup_info_ex.lpAttributeList = attr_list.as_mut_ptr() as _;
```

Comment:
Clippy is mad about this. 

```
`error: casting from `*mut u8` to a more-strictly-aligned pointer (`*mut winapi::um::processthreadsapi::PROC_THREAD_ATTRIBUTE_LIST`)
   --> src\tty\windows\conpty.rs:160:39
    |
160 |     startup_info_ex.lpAttributeList = attr_list.as_mut_ptr() as _;
    |                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: #[deny(clippy::cast_ptr_alignment)] on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#cast_ptr_alignment`
```

Reply:
My understanding is that this warning is occurring because the compiler thinks a `PROC_THREAD_ATTRIBUTE_LIST` contains `*mut c_void` (of size 64 bits) and therefore the struct needs to be aligned to that size, whereas *mut u8 can be aligned to 8 bits. It's not actually an issue as that's just the dummy definition in `winapi-rs` (as the actual allocated size of the struct will be determined by the api call).